File: java_solutions/806.number-of-lines-to-write-string.java
/*
 * @lc app=leetcode id=806 lang=java
 *
 * [806] Number of Lines To Write String
 */

class Solution {
    public int[] numberOfLines(int[] widths, String s) {
        return s.chars().boxed().reduce(new int[]{1, 0}, (state, c) -> widths[c - 97] + state[1] > 100 ? new int[]{state[0] + 1, widths[c - 97]} : new int[]{state[0], state[1] + widths[c - 97]}, (a, b) -> a);
    }
}

File: kotlin_solutions/806.number-of-lines-to-write-string.kt
/*
 * @lc app=leetcode id=806 lang=kotlin
 *
 * [806] Number of Lines To Write String
 */

class Solution { fun numberOfLines(widths: IntArray, s: String): IntArray = s.fold(1 to 0) { (lines, width), c -> (widths[c - 'a'] + width).let { if (it > 100) lines + 1 to widths[c - 'a'] else lines to it } }.let { intArrayOf(it.first, it.second) } }